server/src/core/user/user-service.ts

```diff
@@ -683,6 +684,103 @@ export class Users implements UserService {
     }
   }
 
+  public async addBeneficiary(args: UserAddBeneficiaryArgs): Promise<User> {
+    
+    validators.validatesAddBeneficiary(args);
+    const { phone, name, email } = args;
+    try {
+      const now = new Date();
+      const password = await hashPassword(generatePassword());
+      const user: DbUser = {
+        _id: generateId(),
+        phone,
+        name,
+        password,
+        addedBy: '',
+        isVetted: false,
+        donors: [],
+        roles: ['beneficiary'],
+        createdAt: now,
+        updatedAt: now
+      };
+
+      if (email) {
+        user.email = email;
+      }
+
+      const res = await this.collection.insertOne(user);
+      return getSafeUser(res.ops[0]);
+    }
+    catch(e) {
+      rethrowIfAppError(e);
+
+      if (isMongoDuplicateKeyError(e, args.phone)) {
+        throw createUniquenessFailedError(messages.ERROR_PHONE_ALREADY_IN_USE);
+      }
+
+      if (isMongoDuplicateKeyError(e, args.email)) {
+        throw createUniquenessFailedError(messages.ERROR_EMAIL_ALREADY_IN_USE);
+      }
+
+      throw createDbOpFailedError(e.message);
+    }
+  }
+
+  public async verifyBeneficiaryByName(name: string): Promise<User> {
+    validators.verifyBeneficiaryByName(name);
+    try {
+      const user = await this.verifyBeneficiaryByProperty('name', name);
+      return user;
+    }
+    catch(e) {
+      rethrowIfAppError(e);
+      throw createDbOpFailedError(e.message);
+    }
+  }
+
+  public async verifyBeneficiaryByPhone(phone: string): Promise<User> {
+    validators.verifyBeneficiaryByPhone(phone);
+    try {
+      const user = await this.verifyBeneficiaryByProperty('phone', phone);
+      return user;
+    }
+    catch(e) {
+      rethrowIfAppError(e);
+      throw createDbOpFailedError(e.message);
+    }
+  }
+
+  async verifyBeneficiaryByProperty(property: String, value: String): Promise<User> {
+    let query: any = { roles: { $in: ['beneficiary'] }, isVetted: false };
+
+    if (property === "name") {
+      query = { name: value, ...query };
+    }
+    else if (property === 'phone') {
+      query = { phone: value, ...query };
+    }
+
+    try {
+      const verifiedBeneficiary = await this.collection.findOneAndUpdate(
+        query, 
+        { 
+          $set: { isVetted: true },
+          $currentDate: { updatedAt: true },
+        },
+        { upsert: true, returnOriginal: false }
+      );
```

Comment:
Make sure we only return safe fields (i.e. password should be excluded from the result projection).